File: 347_top_k_frequent_elements/topKFrequent.py
from collections import Counter

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hashmap = {}
        frequency = [[] for _ in range(len(nums) + 1)]

        for num in nums:
            hashmap[num] = hashmap.get(num, 0) + 1
        for i, j in hashmap.items():
            frequency[j].append(i)

        res = []

        for i in range(len(frequency) - 1, 0, -1):
            for j in frequency[i]:
                res.append(j)
                if len(res) == k:
                    return res

        # hashmap = {}
        # res = []

        # for i in nums:
        #     if i in hashmap:
        #         hashmap[i] += 1
        #     else:
        #         hashmap[i] = 1
        # print(hashmap)
        # sorted_hash = list(sorted(hashmap.items(), key=lambda item: item[1]))[::-1]
        # print(sorted_hash)

        # return [i for i, _ in list(sorted_hash)[:k]]
            
        # counter = Counter(nums).most_common(k)
        # return [i for i, _ in list(counter)]

File: 347_top_k_frequent_elements/topKFrequent.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  const hashMap = new Map();
  const freq = new Array(nums.length + 1).fill(null);

  for (let num of nums) {
    if (hashMap.has(num)) {
      const value = hashMap.get(num) + 1;
      hashMap.set(num, value);
    } else {
      hashMap.set(num, 1);
    }
  }

  hashMap.forEach((value, key) => {
    if (Array.isArray(freq[value])) {
      freq[value].push(key);
    } else {
      freq[value] = [key];
    }
  });

  const result = [];

  for (let i = freq.length - 1; i >= 0; i--) {
    if (Array.isArray(freq[i]))
      for (let j of freq[i]) {
        result.push(j);
        if (result.length === k) return result;
      }
  }

  return result;
};
